python/tskit/trees.py

```diff
@@ -10520,6 +10527,112 @@ def sample_nodes_by_ploidy(self, ploidy):
+        and non-sample nodes.
+
+        :param list individuals: Specific individual IDs to include in the VCF. If not
+            specified and the tree sequence contains individuals, all individuals whose
+            nodes all have the flag NODE_IS_SAMPLE set are included.
+        :param int ploidy: The ploidy, or number of nodes per individual. Only used when
+            the tree sequence does not contain individuals. Cannot be used if the tree
+            sequence contains individuals. Defaults to 1 if not specified.
+        :param str name_metadata_key: The key in the individual metadata to use
+            for individual names. Cannot be specified simultaneously with
+            individual_names.
+        :param list individual_names: The names to use for each individual. Cannot
+            be specified simultaneously with name_metadata_key.
+        :return: A VcfModelMapping containing the node-to-individual mapping and
+            individual names.
+        :raises ValueError: If both name_metadata_key and individual_names are specified,
+            if ploidy is specified when individuals are present, if an invalid individual
+            ID is specified, if a specified individual has no nodes, or if the number of
+            individuals doesn't match the number of names.
+        """
+        if include_non_sample_nodes is None:
+            include_non_sample_nodes = False
+
+        if name_metadata_key is not None and individual_names is not None:
+            raise ValueError(
+                "Cannot specify both name_metadata_key and individual_names"
+            )
+
+        if self.num_individuals > 0 and ploidy is not None:
+            raise ValueError(
+                "Cannot specify ploidy when individuals are present in the tree sequence"
+            )
+
+        if self.num_individuals > 0 and np.any(
+            np.logical_and(
+                self.nodes_individual == tskit.NULL,
+                self.nodes_flags & tskit.NODE_IS_SAMPLE,
+            )
+        ):
+            warnings.warn(
+                "At least one sample node does not have an individual ID.", stacklevel=1
+            )
+
+        if self.num_individuals == 0 and individuals is None:
+            if ploidy is None:
+                ploidy = 1
+            individuals_nodes = self.sample_nodes_by_ploidy(ploidy)
+            if individual_names is None:
+                individual_names = [f"tsk_{i}" for i in range(len(individuals_nodes))]
+        else:
+            if individuals is None:
+                individuals = np.arange(self.num_individuals, dtype=np.int32)
+            if min(individuals) < 0 or max(individuals) >= self.num_individuals:
+                raise ValueError("Invalid individual ID")
+
+            individuals_nodes = self.individuals_nodes[individuals]
+            if not include_non_sample_nodes:
+                individuals_nodes[
+                    np.logical_not(
```

Comment:
This is a bit of a corner case, so I think we can just do a loop to make our lives a bit easier:
```
non_sample_nodes =  np.logical_not(self.nodes_flags[individuals_nodes] & tskit.NODE_IS_SAMPLE)
if np.any(non_sample_nodes) and not include_non_sample_nodes:
    individuals_nodes[non_sample_nodes] = -1
    for index in np.where(non_sample_nodes):
        # do the switcharoo to make sure -1 is at the end
```